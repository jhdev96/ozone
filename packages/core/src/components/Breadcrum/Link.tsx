import React from 'react';
import { Arrow } from 'icons';
import classNames from '../../utils/classNames';
import Colors from 'colors';
import styles from './index.module.scss';


export interface LinkProps {
  /** 
   * Indicates visually whether the path is active. 
   */
  isActive?: boolean;
  /** 
   * Function that is called when a path is clicked. 
   */
  onClick?: () => void;
  /**
   * A selector used in testing environments.
   */
  dataTestId?: string;
  /** 
   * Content to render inside a path. 
   */
  children?: React.ReactNode;
}

export default function Link({ 
  isActive = false, 
  onClick,
  dataTestId,
  children
}: LinkProps): JSX.Element 
{
  return (
    <>
      <button
        onClick={onClick} 
        className={classNames(
          styles.link, 
          isActive && styles.active
        )} 
        data-testid={dataTestId}
      >
        {children}
      </button>
      {!isActive && (
        <Arrow 
          direction="right" 
          color={Colors.grayScale.GRAY3} 
        />
      )}
    </>
  )
};

Link.displayName = 'BreadcrumLink';