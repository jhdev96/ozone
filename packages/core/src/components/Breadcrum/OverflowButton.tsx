import React from 'react';
import classNames from '../../utils/classNames';
import { ThemeType } from '../../utils/theme';
import styles from './index.module.scss';


interface OverflowButtonProps {
  /**
   * A light or dark theme for the button. 
   */
  theme?: ThemeType;
  /**
   * If true, gives the button element
   * a css class to visually indicate
   * that it is currently selected.
   */
  selected?: boolean;
  /** 
   * Aria label for the overflow button. 
   */
  ariaLabel?: string;
  /**
   * A function that fires when the button
   * is clicked.
   */
  onClick?: () => void;
}

export default function OverflowButton({
  theme = "light",
  selected,
  ariaLabel,
  onClick
}: OverflowButtonProps): JSX.Element 
{
  return (
    <button 
      className={classNames(
        styles.dots, 
        styles[theme],
        selected && styles.selected
      )} 
      onClick={onClick}
      aria-label={ariaLabel}
    >
      <div className={styles.dot} />
      <div className={styles.dot} />
      <div className={styles.dot} />
    </button> 
  )
}

OverflowButton.displayName = 'OverflowButton';