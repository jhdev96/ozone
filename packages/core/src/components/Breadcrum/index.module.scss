@import '../../scss/root';


$borderRadius: 6px;

.breadcrum { display: inline-flex; }
 
.link {
  padding: 12px;
  color: var(--blue-core);
  background-color: transparent;
  border: 0 !important;
  border-radius: $borderRadius;
  font-size: 0.95rem;
  cursor: pointer !important;

  &:not(.active) {
    &:hover {
      background-color: var(--gray-0);
    }
  }
}

.overflow {
  display: inline-flex;
  position: relative;

  .dots {
    @extend .link;

    border-radius: $borderRadius;
    transition: 0.2s background-color;
    cursor: pointer;
  }

  .dot {
    display: inline-flex;
    margin-right: 0.1rem;
    margin-left: 0.1rem;
    margin-bottom: 0.2rem;
    width: 4px;
    height: 4px;
    background-color: var(--blue-core);
    border-radius: 50%;
  }

  .menu {
    position: absolute;
    z-index: 1000;
    top: 30.1px;
    min-width: 150px;
    visibility: hidden;
    border-radius: $borderRadius;
    opacity: 0;
    box-shadow: 1px 5px 15px 2px rgba(0, 0, 0, 0.077);

    svg { display: none; }

    :first-child {
      border-radius: $borderRadius $borderRadius 0 0 !important;
    }

    :nth-last-of-type(1) {
      border-radius: 0 0 $borderRadius $borderRadius !important;
    }

    .link {
      width: 100%;
      display: block !important;
      padding: 10px !important;
      text-align: left;
      border-radius: 0;

      &:hover {
        background-color: var(--blue-0);
      }
    }
  }

  .visible {
    visibility: visible !important;
    transform: translateY(11px);
    opacity: 1 !important;
    transition: 0.3s opacity, 0.3s transform;
  }

  .openUp { transform: translateY(-11px); }
}

/* Modifiers for props */
.active {
  color: var(--gray-4) !important;
  cursor: default !important;
}

.selected {
  background-color: var(--blue-0) !important;

  .dots:hover {
    background-color: var(--blue-0) !important;
  }
}

.focus { outline: 1px solid #000000; }
.light { background-color: #fff; }
.dark { background-color: var(--black-4); }