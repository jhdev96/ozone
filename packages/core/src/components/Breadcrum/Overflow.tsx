import React, { useState, useRef, useEffect } from 'react';
import { Arrow } from 'icons';
import ClickManager from '../ClickManager';
import Colors from 'colors';
import { ThemeType } from '../../utils/theme';
import classNames from '../../utils/classNames';
import styles from './index.module.scss';
import OverflowManager from '../OverflowManager';
import OverflowButton from './OverflowButton';


interface OverflowProps {
  /** 
   * A light or dark theme for the hidden menu. 
   */
  theme?: ThemeType
  /** 
   * Aria label for the overflow button. 
   */ 
  ariaLabel?: string;
  /**
   * A selector used in testing environments.
   * 
   * This should ONLY be used in a testing 
   * environment; otherwise, the <OverflowManager />
   * component will not work as intended.
   */
  dataTestId?: string;
  /** 
   * Link elements to be displayed in a hidden menu. 
   */
  children: React.ReactNode[]
}

export default function Overflow({
  theme = "light",
  ariaLabel,
  dataTestId,
  children
}: OverflowProps): JSX.Element 
{
  const [visible, setVisible] = useState(false);
  const [aboveTarget, setAboveTarget] = useState(false);
  const divRef = useRef<HTMLDivElement>(null);
  const menuRef = useRef<HTMLDivElement>(null);
  const MENU_OFFSET = 11; // For transition effect

  const positionMenu = (menu: HTMLElement | undefined | null) => {
    if (!menu) return;
    
    menu.style.top = `-${menu.clientHeight - MENU_OFFSET}px`;
    setAboveTarget(true);
  }

  const toggleVisibility = () => {
    setVisible(!visible);

    if (menuRef.current) {
      menuRef.current.style.top = '';
    }
  }

  const handleOutsideClick = () => {
    visible && setVisible(false);
  }

  useEffect(() => {
    if (!visible) setAboveTarget(false);
  }, [visible]);

  return (
    <ClickManager ref={divRef} onClick={handleOutsideClick}>
      <OverflowManager 
        ref={menuRef} 
        isVisible={!dataTestId && visible}
        onOverflow={() => positionMenu(menuRef.current)} 
        onVisible={() => setAboveTarget(false)}
        updateParameters={[visible]}
      >
        <span 
          className={styles.overflow} 
          data-testid={dataTestId}
        >
          <OverflowButton 
            selected={visible}
            theme={theme}
            onClick={toggleVisibility}
            ariaLabel={ariaLabel}
          />
          <Arrow 
            direction="right" 
            color={Colors.grayScale.GRAY3} 
          />
          <div 
            className={classNames(
              styles.menu, 
              styles[theme],
              visible     && styles.visible,
              aboveTarget && styles.openUp
            )}
            ref={menuRef}
          >
            {children}
          </div>
        </span>
      </OverflowManager>
    </ClickManager>
  )
}

Overflow.displayName = 'BreadcrumOverflow';