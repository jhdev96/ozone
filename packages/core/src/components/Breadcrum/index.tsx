import React from 'react';
import Link from './Link';
import Overflow from './Overflow';
import styles from './index.module.scss';


interface BreadcrumProps {
  /** 
   * A selector used in testing environments. 
   */
  dataTestId?: string;
  /** 
   * Content to render inside a path. 
   */
  children?: React.ReactNode | React.ReactNode[];
}

export default function Breadcrum({
  dataTestId,
  children
}: BreadcrumProps): JSX.Element 
{
  return (
    <nav 
      className={styles.breadcrum} 
      data-testid={dataTestId}
    >
      {children}
    </nav>
  )
}

Breadcrum.Overflow = Overflow;
Breadcrum.Link = Link;
Breadcrum.displayName = 'Breadcrum';