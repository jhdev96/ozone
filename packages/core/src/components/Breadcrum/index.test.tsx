import React from 'react';
import { render, screen, fireEvent } from '@testing-library/react';
import Breadcrum from '.';


describe('<Breadcrum />', () => {
  it('renders a breadcrum with 3 links', () => {
    const {container} = render(
      <Breadcrum dataTestId="test">
        <Breadcrum.Link>Link 1</Breadcrum.Link>
        <Breadcrum.Link>Link 2</Breadcrum.Link>
        <Breadcrum.Link isActive>Link 3</Breadcrum.Link>
      </Breadcrum>
    );

    expect(container.querySelectorAll('[data-testid="test"] .link'))
      .toHaveLength(3);

    expect(container).toMatchSnapshot();
  });

  it('calls `onClick` when a link is clicked', () => {
    const onClick1 = jest.fn();
    const onClick2 = jest.fn();

    render(
      <Breadcrum dataTestId="test">
        <Breadcrum.Link 
          onClick={onClick1}
          dataTestId="link1"
        >
          Link 1
        </Breadcrum.Link>
        <Breadcrum.Link 
          onClick={onClick2}
          dataTestId="link2"
        >
          Link 2
        </Breadcrum.Link>
        <Breadcrum.Link isActive>Link 3</Breadcrum.Link>
      </Breadcrum>
    );

    fireEvent.click(screen.getByTestId('link1'));
    fireEvent.click(screen.getByTestId('link2'));

    expect(onClick1).toBeCalled();
    expect(onClick2).toBeCalled();
  });

  it('displays the links when Breadcrum.Overflow is clicked', () => {
    render(
      <Breadcrum>
        <Breadcrum.Overflow dataTestId="test">
          <Breadcrum.Link>Root</Breadcrum.Link>
          <Breadcrum.Link>Folder1</Breadcrum.Link>
          <Breadcrum.Link>Folder2</Breadcrum.Link>
        </Breadcrum.Overflow>
      </Breadcrum>
    );

    const overflow = screen.getByTestId('test');

    expect(overflow.querySelector('.menu'))
      .not.toHaveClass('visible');

    // clicks the overflow button with the 3 dots.
    fireEvent.click(overflow.querySelector('.dots')!);

    expect(overflow.querySelector('.menu'))
      .toHaveClass('visible');
  });

  it('renders a breadcrum with a dark theme', () => {
    render(
      <Breadcrum dataTestId="test">
        <Breadcrum.Link>Home</Breadcrum.Link>
        <Breadcrum.Overflow theme="dark">
          <Breadcrum.Link>Documents</Breadcrum.Link>
          <Breadcrum.Link>Shared</Breadcrum.Link>
          <Breadcrum.Link>Downloads</Breadcrum.Link>
        </Breadcrum.Overflow>
        <Breadcrum.Link>Folder1</Breadcrum.Link>
        <Breadcrum.Link isActive>Folder2</Breadcrum.Link>
      </Breadcrum>
    );

    expect(screen.getByTestId('test')).toMatchSnapshot();
  });
});