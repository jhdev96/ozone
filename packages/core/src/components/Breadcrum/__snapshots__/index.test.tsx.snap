// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Breadcrum /> renders a breadcrum with 3 links 1`] = `
<div>
  <nav
    class="breadcrum"
    data-testid="test"
  >
    <button
      class="link"
    >
      Link 1
    </button>
    <svg
      height="9.1"
      style="transform: rotate(90deg); align-self: center;"
      viewBox="0 0 14.4 9.1"
      width="10.4"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M7.2,0,0,7.011,2.145,9.1,7.2,4.177,12.255,9.1,14.4,7.011Z"
        fill="#C4C5D0"
        id="Path_36"
        transform="translate(0 0)"
      />
    </svg>
    <button
      class="link"
    >
      Link 2
    </button>
    <svg
      height="9.1"
      style="transform: rotate(90deg); align-self: center;"
      viewBox="0 0 14.4 9.1"
      width="10.4"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M7.2,0,0,7.011,2.145,9.1,7.2,4.177,12.255,9.1,14.4,7.011Z"
        fill="#C4C5D0"
        id="Path_36"
        transform="translate(0 0)"
      />
    </svg>
    <button
      class="link active"
    >
      Link 3
    </button>
  </nav>
</div>
`;

exports[`<Breadcrum /> renders a breadcrum with a dark theme 1`] = `
<nav
  class="breadcrum"
  data-testid="test"
>
  <button
    class="link"
  >
    Home
  </button>
  <svg
    height="9.1"
    style="transform: rotate(90deg); align-self: center;"
    viewBox="0 0 14.4 9.1"
    width="10.4"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      d="M7.2,0,0,7.011,2.145,9.1,7.2,4.177,12.255,9.1,14.4,7.011Z"
      fill="#C4C5D0"
      id="Path_36"
      transform="translate(0 0)"
    />
  </svg>
  <div>
    <div>
      <span
        class="overflow"
      >
        <button
          class="dots dark"
        >
          <div
            class="dot"
          />
          <div
            class="dot"
          />
          <div
            class="dot"
          />
        </button>
        <svg
          height="9.1"
          style="transform: rotate(90deg); align-self: center;"
          viewBox="0 0 14.4 9.1"
          width="10.4"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M7.2,0,0,7.011,2.145,9.1,7.2,4.177,12.255,9.1,14.4,7.011Z"
            fill="#C4C5D0"
            id="Path_36"
            transform="translate(0 0)"
          />
        </svg>
        <div
          class="menu dark"
        >
          <button
            class="link"
          >
            Documents
          </button>
          <svg
            height="9.1"
            style="transform: rotate(90deg); align-self: center;"
            viewBox="0 0 14.4 9.1"
            width="10.4"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M7.2,0,0,7.011,2.145,9.1,7.2,4.177,12.255,9.1,14.4,7.011Z"
              fill="#C4C5D0"
              id="Path_36"
              transform="translate(0 0)"
            />
          </svg>
          <button
            class="link"
          >
            Shared
          </button>
          <svg
            height="9.1"
            style="transform: rotate(90deg); align-self: center;"
            viewBox="0 0 14.4 9.1"
            width="10.4"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M7.2,0,0,7.011,2.145,9.1,7.2,4.177,12.255,9.1,14.4,7.011Z"
              fill="#C4C5D0"
              id="Path_36"
              transform="translate(0 0)"
            />
          </svg>
          <button
            class="link"
          >
            Downloads
          </button>
          <svg
            height="9.1"
            style="transform: rotate(90deg); align-self: center;"
            viewBox="0 0 14.4 9.1"
            width="10.4"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M7.2,0,0,7.011,2.145,9.1,7.2,4.177,12.255,9.1,14.4,7.011Z"
              fill="#C4C5D0"
              id="Path_36"
              transform="translate(0 0)"
            />
          </svg>
        </div>
      </span>
    </div>
  </div>
  <button
    class="link"
  >
    Folder1
  </button>
  <svg
    height="9.1"
    style="transform: rotate(90deg); align-self: center;"
    viewBox="0 0 14.4 9.1"
    width="10.4"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      d="M7.2,0,0,7.011,2.145,9.1,7.2,4.177,12.255,9.1,14.4,7.011Z"
      fill="#C4C5D0"
      id="Path_36"
      transform="translate(0 0)"
    />
  </svg>
  <button
    class="link active"
  >
    Folder2
  </button>
</nav>
`;
